lib/puppet/provider/mysql_grant/mysql.rb

```diff
@@ -129,11 +131,28 @@ def flush
 
   mk_resource_methods
 
+  def diff_privileges(privileges_old, privileges_new)
+    diff = {:revoke => Array.new, :grant => Array.new}
+    if privileges_new.include? 'ALL' or privileges_old.include? 'ALL'
+      diff[:revoke] = privileges_old
+      diff[:grant] = privileges_new
+    else
+      diff[:revoke] = privileges_old - privileges_new
+      diff[:grant] = privileges_new - privileges_old
+    end
+    return diff
+  end
+
   def privileges=(privileges)
-    revoke(@property_hash[:user], @property_hash[:table])
-    grant(@property_hash[:user], @property_hash[:table], privileges, @property_hash[:options])
+    diff = diff_privileges(@property_hash[:privileges], privileges)
+    if not diff[:revoke].empty?
+      revoke(@property_hash[:user], @property_hash[:table], diff[:revoke])
+    end
+    if not diff[:grant].empty?
+      grant(@property_hash[:user], @property_hash[:table], diff[:grant], @property_hash[:options])
+    end
     @property_hash[:privileges] = privileges
-
+    mysql([defaults_file, '-NBe', 'FLUSH PRIVILEGES'].compact)
```

Comment:
Flush is automatically called by the puppet transaction after the resource
is finished.
On May 2, 2015 1:44 PM, "Artur Gadelshin" <notifications@github.com> wrote:

> In lib/puppet/provider/mysql_grant/mysql.rb
> <https://github.com/puppetlabs/puppetlabs-mysql/pull/703#discussion_r29547221>
> :
>
> >      @property_hash[:privileges] = privileges
> > -
> > +    mysql([defaults_file, '-NBe', 'FLUSH PRIVILEGES'].compact)
>
> Yes, there is flush method, but I can't find place in module, where this
> method is called.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/puppetlabs-mysql/pull/703/files#r29547221>.
>